### ISSUE TITLE
Fixing UGV performance with SubT stack. UGV keep toppling on stairs. 

### ISSUE THREAD

**Issue by okyspace** [DinoHub/devmgmt-monkeybotics#115]

**Comment by timothymah** [comment]
Kevin Hostler mentioned he will get back to us on this when we met him on 3 Nov 2021

**Comment by Blqe** [comment]
Check on timestep. Real or Sim

**Comment by jehontan** [comment]
Elaborating on the use of sim time:

## Intuition

Toppling after some time, but not immediately failing on launch, is an indication of the controllers becoming unstable. If the controllers were previously tuned for the robot configuration tested, it is likely that the controller parameters are set correctly and that something about the simulation setup is wrong.

Instability could be caused by the control code running at a different rate relative to the physical simulation compared to  when the control parameters were originally tuned. It could also be caused by computer performance (OS scheduler, context switches) causing uneven durations (system time) in between simulation timesteps (could be constant in sim time).

## ROS Configuration Issues

By default the ROS2 client libraries use system time (`SystemTime`). A node has to be explictly told to use simulation time (`ROSTime`) by setting the `use_sim_time` parameter to `True`. However `ROSTime` defaults to an alias for `SystemTime`, unless the `/clock` topic is present, and even if present `ROSTime` will default to 0 is `/clock` is not published to. See [this](https://design.ros2.org/articles/clock_and_time.html) for details.

Gazebo's ROS2  integration does not automatically publish `/clock` (see this [issue](https://github.com/ros2/rclcpp/issues/940)). Importantly, it is necessary to start `gzserver` with the `gazebo_ros_init` plugin (see this [article](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-ROS-Clocks-and-sim-time)) in order for the `/clock` topic to be published to.

## TL;DR

Checking if you're using sim time involves the following:

- [ ] Check that nodes have the `use_sim_time` parameter set to `True`, either in explicit cmd line parameters or associated YAML files
- [ ] Check that`gzserver` is launched with the `gazebo_ros_init` plugin, i.e. `$ gzserver -s gazebo_ros_init`
- [ ] Verify that the `/clock` topic is present and echo it to verify that the values are being updated.

**Comment by amokduke** [comment]
Hey @jehontan,

Thank you for helping to look at this. =)

Previously I asked Kat about the simulation "overrunning" too:
![image.png](https://images.zenhubusercontent.com/60af601746d84465c9496341/fe7c393c-fcd9-4489-9d9d-53e3d5d1c2e0)

We didn't have time to follow up on it and when we got "better machines", the number of "weird behaviours" from UAVs went down. However, for UGVs, the toppling still occurs. The UGVs requires less computation power than UAVs so if the machine can handle UAVs.... by right computer performance "should not" be the reason behind the toppling....  what do you think?

**Comment by jonyktan** [comment]
@timothymah are you able to work on this for this sprint? If not propose to shift to Product Backlog

**Comment by jehontan** [comment]
Sorry for the late reply.

@amokduke Why would you say the UGV requires less computation than tthe UAV?

The computational demands may not be high, but the control loop of the UGV might be more sensitive to variation in sampling time, especially for more dynamic movement like stairs.

**Comment by amokduke** [comment]
> Sorry for the late reply.
> 
> @amokduke Why would you say the UGV requires less computation than tthe UAV?
> 
> The computational demands may not be high, but the control loop of the UGV might be more sensitive to variation in sampling time, especially for more dynamic movement like stairs.

Hi @jehontan,

No worries, =) We used the built-in Ubuntu’s system monitor UI to check the CPU usage when we ran the scenarios then. The wheeled UGVs were using less computational resources. 

At the moment, we don't have any legged UGVs into our stack yet. With legged UGVs, yeah, I think the UGV may take more computational resources when climbing stairs too.

**Comment by jehontan** [comment]
Yea for sure the wheeled UGVs will be less intensive as they won't be running any fast controllers for balancing. The legged UGV base computational requirement even for level walking is likely to exceed the quadcopters.
